### PR TITLE
WIP: using multistream muxer

### PR DESCRIPTION
**Pull request by whyrusleeping**
This PR adds multistream as the protocol muxer in go-ipfs and allows the selection of yamux or spdystream for stream muxing. It should be merged on top of the dev0.4.0 branch.

Currently, some tests are failing, i've looked at them briefly, but theyre out of my domain of knowledge. if @jbenet could take a look, that would be great.

License: MIT
Signed-off-by: Jeromy <jeromyj@gmail.com>

### CODE REVIEW

**Comment by jbenet** [comment]
@whyrusleeping code LGTM but the tests seem to be deadlocking. maybe things not getting added to the muxer right? or negotiated by the muxer right?

**Comment by whyrusleeping** [comment]
where are you seeing a deadlock? I see some tests failing locally, but nothing seems like its actually hung

**Comment by jbenet** [comment]
the travis tests hang

**Comment by whyrusleeping** [comment]
(also, If you could look at the test failures in `p2p/protocol/identify` that would be great)

**Comment by GitCop** [comment]
There were the following issues with your Pull Request

 * Commit: 41e570172d8e77213739bc4e8dbb55ce6fc43011
  * Invalid signoff. Commit message must end with **License: MIT
Signed-off-by: .* <.*>**

 * Commit: eb42e2f0bd9a4fc22ec23d4b4a43863fa0251317
  * Invalid signoff. Commit message must end with **License: MIT
Signed-off-by: .* <.*>**

 * Commit: 12a44574180edf412f65528a72eced0b14bd3cad
  * Invalid signoff. Commit message must end with **License: MIT
Signed-off-by: .* <.*>**

 * Commit: 27dc7c515b598c9b42cf3fc73977d30b116b6588
  * Invalid signoff. Commit message must end with **License: MIT
Signed-off-by: .* <.*>**

 * Commit: 4aee752d963c8f664593a489b5e5df3f42591f45
  * Invalid signoff. Commit message must end with **License: MIT
Signed-off-by: .* <.*>**

 * Commit: 7f6e78acc4cfa28f31fb97dce6bc2ab474fefb43
  * Invalid signoff. Commit message must end with **License: MIT
Signed-off-by: .* <.*>**

 * Commit: 474d041604e603040d6d0594b75504ec5e0ae9a7
  * Invalid signoff. Commit message must end with **License: MIT
Signed-off-by: .* <.*>**

 * Commit: 19658648c857ea7fdd72c471a2d66997279ee9db
  * Invalid signoff. Commit message must end with **License: MIT
Signed-off-by: .* <.*>**

 * Commit: d1f484042240692bc1e6000af0663c8c4aabde9e
  * Invalid signoff. Commit message must end with **License: MIT
Signed-off-by: .* <.*>**

 * Commit: 053a7ee3279f37081a25ec35dbe5e53654e388e4
  * Invalid signoff. Commit message must end with **License: MIT
Signed-off-by: .* <.*>**

 * Commit: 3e08d5e9076be876afeecf1b3595c90f22176f6e
  * Invalid signoff. Commit message must end with **License: MIT
Signed-off-by: .* <.*>**

 * Commit: aa2b9fe127731f8828075183606810ca722f5043
  * Invalid signoff. Commit message must end with **License: MIT
Signed-off-by: .* <.*>**

 * Commit: 3bdf75df82bc3b05427d64286f9af57899d4785b
  * Invalid signoff. Commit message must end with **License: MIT
Signed-off-by: .* <.*>**

 * Commit: 42b9e53b5fe5c44b32bba327df791b8355b7fcf9
  * Invalid signoff. Commit message must end with **License: MIT
Signed-off-by: .* <.*>**

 * Commit: 86776db140d214b1edf2d89bf5f54a4679850f1e
  * Invalid signoff. Commit message must end with **License: MIT
Signed-off-by: .* <.*>**

 * Commit: 1493ca0ec45252b35ef67d3f19730284e469b0eb
  * Invalid signoff. Commit message must end with **License: MIT
Signed-off-by: .* <.*>**

 * Commit: e72dbf333fc6d0677168dd766be6afedaaee52ed
  * Invalid signoff. Commit message must end with **License: MIT
Signed-off-by: .* <.*>**

 * Commit: 617714a759960d7bd07de71722efb84793e722df
  * Invalid signoff. Commit message must end with **License: MIT
Signed-off-by: .* <.*>**

 * Commit: c42b501f7e0a2e2289353a3611d49acf3f8a2834
  * Invalid signoff. Commit message must end with **License: MIT
Signed-off-by: .* <.*>**

 * Commit: 956e398768b2e28f2dbd565efe3dbb56d4f1fa15
  * Invalid signoff. Commit message must end with **License: MIT
Signed-off-by: .* <.*>**

[Guidelines](https://github.com/ipfs/community/blob/master/contribution-guidelines.md#commit-messages) are available to help. Your feedback on GitCop is welcome on [this issue](https://github.com/ipfs/infrastructure/issues/23)

---
This message was auto-generated by https://gitcop.com

**Comment by whyrusleeping** [comment]
These test failures are rather frustrating... nothing appears to be concretely related... although *every* travis test failed on that set.

**Comment by jbenet** [comment]
@whyrusleeping all circlecis failed on this:

```
not ok 7 - transport should be unencrypted
#	
#	  go-sleep 0.5s | nc localhost "$PORT_SWARM" >swarmnc &&
#	  test_must_fail grep -q "AES-256,AES-128" swarmnc &&
#	  grep -q "/ipfs/identify" swarmnc ||
#	  test_fsh cat swarmnc
#	

expecting success: 
```

**Comment by whyrusleeping** [comment]
ah, didnt catch that one. That test fails no matter what on my machine anyways so i dont run it locally, speaking of which, i wrote this to do it better: https://github.com/whyrusleeping/ngrep

**Comment by GitCop** [comment]
There were the following issues with your Pull Request

 * Commit: 41e570172d8e77213739bc4e8dbb55ce6fc43011
  * Invalid signoff. Commit message must end with **License: MIT
Signed-off-by: .* <.*>**

 * Commit: eb42e2f0bd9a4fc22ec23d4b4a43863fa0251317
  * Invalid signoff. Commit message must end with **License: MIT
Signed-off-by: .* <.*>**

 * Commit: 12a44574180edf412f65528a72eced0b14bd3cad
  * Invalid signoff. Commit message must end with **License: MIT
Signed-off-by: .* <.*>**

 * Commit: 27dc7c515b598c9b42cf3fc73977d30b116b6588
  * Invalid signoff. Commit message must end with **License: MIT
Signed-off-by: .* <.*>**

 * Commit: 4aee752d963c8f664593a489b5e5df3f42591f45
  * Invalid signoff. Commit message must end with **License: MIT
Signed-off-by: .* <.*>**

 * Commit: 7f6e78acc4cfa28f31fb97dce6bc2ab474fefb43
  * Invalid signoff. Commit message must end with **License: MIT
Signed-off-by: .* <.*>**

 * Commit: 474d041604e603040d6d0594b75504ec5e0ae9a7
  * Invalid signoff. Commit message must end with **License: MIT
Signed-off-by: .* <.*>**

 * Commit: 19658648c857ea7fdd72c471a2d66997279ee9db
  * Invalid signoff. Commit message must end with **License: MIT
Signed-off-by: .* <.*>**

 * Commit: d1f484042240692bc1e6000af0663c8c4aabde9e
  * Invalid signoff. Commit message must end with **License: MIT
Signed-off-by: .* <.*>**

 * Commit: 053a7ee3279f37081a25ec35dbe5e53654e388e4
  * Invalid signoff. Commit message must end with **License: MIT
Signed-off-by: .* <.*>**

 * Commit: 3e08d5e9076be876afeecf1b3595c90f22176f6e
  * Invalid signoff. Commit message must end with **License: MIT
Signed-off-by: .* <.*>**

 * Commit: aa2b9fe127731f8828075183606810ca722f5043
  * Invalid signoff. Commit message must end with **License: MIT
Signed-off-by: .* <.*>**

 * Commit: 3bdf75df82bc3b05427d64286f9af57899d4785b
  * Invalid signoff. Commit message must end with **License: MIT
Signed-off-by: .* <.*>**

 * Commit: 42b9e53b5fe5c44b32bba327df791b8355b7fcf9
  * Invalid signoff. Commit message must end with **License: MIT
Signed-off-by: .* <.*>**

 * Commit: 86776db140d214b1edf2d89bf5f54a4679850f1e
  * Invalid signoff. Commit message must end with **License: MIT
Signed-off-by: .* <.*>**

 * Commit: 1493ca0ec45252b35ef67d3f19730284e469b0eb
  * Invalid signoff. Commit message must end with **License: MIT
Signed-off-by: .* <.*>**

 * Commit: e72dbf333fc6d0677168dd766be6afedaaee52ed
  * Invalid signoff. Commit message must end with **License: MIT
Signed-off-by: .* <.*>**

 * Commit: 617714a759960d7bd07de71722efb84793e722df
  * Invalid signoff. Commit message must end with **License: MIT
Signed-off-by: .* <.*>**

 * Commit: c42b501f7e0a2e2289353a3611d49acf3f8a2834
  * Invalid signoff. Commit message must end with **License: MIT
Signed-off-by: .* <.*>**

 * Commit: 956e398768b2e28f2dbd565efe3dbb56d4f1fa15
  * Invalid signoff. Commit message must end with **License: MIT
Signed-off-by: .* <.*>**

[Guidelines](https://github.com/ipfs/community/blob/master/contribution-guidelines.md#commit-messages) are available to help. Your feedback on GitCop is welcome on [this issue](https://github.com/ipfs/infrastructure/issues/23)

---
This message was auto-generated by https://gitcop.com

**Comment by whyrusleeping** [comment]
look at all that green.

File: pin/pin.go [hunk 1]
```diff
@@ -243,7 +244,9 @@ func LoadPinner(d ds.ThreadSafeDatastore, dserv mdag.DAGService) (Pinner, error)
 
 	rootKey := key.Key(rootKeyBytes)
 
-	ctx := context.TODO()
+	ctx, cancel := context.WithTimeout(context.TODO(), time.Second*5)
+	defer cancel()
```
Comment (jbenet) [inline pin/pin.go]: so this causes exiting on error, correct?

**Comment by GitCop** [comment]
There were the following issues with your Pull Request

 * Commit: ef347684bdb626f2234323336d51111e8858333a
  * Invalid signoff. Commit message must end with **License: MIT
Signed-off-by: .* <.*>**

[Guidelines](https://github.com/ipfs/community/blob/master/contribution-guidelines.md#commit-messages) are available to help. Your feedback on GitCop is welcome on [this issue](https://github.com/ipfs/infrastructure/issues/23)

---
This message was auto-generated by https://gitcop.com

**Comment by jbenet** [comment]
squashed altogether for tests. (previous head: 2474247)

**Pull request merged**